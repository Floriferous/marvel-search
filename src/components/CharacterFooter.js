import React from 'react';
import PropTypes from 'prop-types';

const CharacterFooter = ({ name }) => (
  <div className="footer">
    <h3>{name}</h3>
  </div>
);

CharacterFooter.propTypes = {
  name: PropTypes.string.isRequired,
};

export default CharacterFooter;
