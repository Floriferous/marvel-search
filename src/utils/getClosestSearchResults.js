import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
