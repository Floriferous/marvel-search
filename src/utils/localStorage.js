export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
