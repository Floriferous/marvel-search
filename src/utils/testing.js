import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
