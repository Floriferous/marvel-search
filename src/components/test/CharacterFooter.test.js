import React from 'react';
import renderer from 'react-test-renderer';

import { testComponent } from '../../utils/testing';
import CharacterFooter from '../CharacterFooter';

describe('CharacterFooter', () => {
  let props;
  const component = () => testComponent(CharacterFooter, props);

  beforeEach(() => {
    props = { name: 'batman' };
    testComponent.reset();
  });

  it('matches its snapshot', () => {
    const tree = renderer.create(<CharacterFooter />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a div with class footer', () => {
    expect(component().find('div').length).toBe(1);
    expect(component()
      .find('div')
      .hasClass('footer')).toBe(true);
  });

  describe('the div', () => {
    let div;

    beforeEach(() => {
      div = component()
        .find('div')
        .shallow();
    });

    it('contains an h3 with name in it', () => {
      expect(div.find('h3').length).toBe(1);
      expect(div.find('h3').text()).toBe(props.name);
    });
  });
});
