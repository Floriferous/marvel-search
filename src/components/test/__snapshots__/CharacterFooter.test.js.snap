// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterFooter matches its snapshot 1`] = `
<div
  className="footer"
>
  <h3 />
</div>
`;
